x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,339 @@
+            if (ignoreMalformed != null) {
+                return new Explicit<>(ignoreMalformed, true);
+            }
+            if (context.indexSettings() != null) {
+                return new Explicit<>(IGNORE_MALFORMED_SETTING.get(context.indexSettings()), false);
+            }
+            return AggregateMetricFieldMapper.Defaults.IGNORE_MALFORMED;
+        }
+
+        public AggregateMetricFieldMapper.Builder metrics(Set<String> metrics) {
+            this.metrics = metrics;
+            return builder;
+        }
+
+        protected Explicit<Set<String>> metrics(BuilderContext context) {
+            if (metrics != null) {
+                return new Explicit<>(metrics, true);
+            }
+            return Defaults.METRICS;
+        }
+
+        @Override
+        public AggregateMetricFieldMapper build(BuilderContext context) {
+            setupFieldType(context);
+            return new AggregateMetricFieldMapper(name, fieldType, defaultFieldType,
+                context.indexSettings(), multiFieldsBuilder.build(this, context),
+                ignoreMalformed(context), metrics(context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<Builder, AggregateMetricFieldMapper> parse(String name,
+                                                                         Map<String, Object> node,
+                                                                         ParserContext parserContext) throws MapperParsingException {
+            AggregateMetricFieldMapper.Builder builder = new AggregateMetricFieldMapper.Builder(name);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext(); ) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals(Names.METRICS.getPreferredName())) {
+                    String metricsStr[] = XContentMapValues.nodeStringArrayValue(propNode);
+                    // Convert the array of Metric enum to a Set of their string representation
+                    Set<String> supporteddMetrics = Arrays.stream(Metrics.values()).map(s -> s.toString()).collect(Collectors.toSet());
+
+                    // Make sure that metrics are supported
+                    Set<String> parsedMetrics = new LinkedHashSet<>();
+                    for (int i = 0; i < metricsStr.length; i++) {
+                        if (supporteddMetrics.contains(metricsStr[i]) == false) {
+                            throw new MapperParsingException("Metric [" + metricsStr[i] + "] is not supported.");
+                        } else {
+                            parsedMetrics.add(metricsStr[i]);
+                        }
+                    }
+                    builder.metrics(parsedMetrics);
```

Comment:
And if we do end up creating a set of enums, we'd need to change the builder (and all downstream users) to accept the enum instead of strings.

Reply:
Fixed